### PR TITLE
Add ID Fetcher in Entrytypedialog

### PR DESCRIPTION
**Pull request by zellerdev**
Add an extra field in the EntryTypeDialog to fetch the ID Based fetcher. 
The field is in on the bottom of the dialog and is named `ID based generator`.

- [ ] DoiFetcher merged #1885 
- [ ] AdsFetcher merged #1923 
- [x] Change in CHANGELOG.md described
- [ ] Tests created for changes
- [x] Screenshots added (for bigger UI changes)
- [x] Manually tested changed features in running JabRef

Normal
dialog30-f

Searching
dialog31-f

No internet connection
dialog34-f

Error message
fetcher35-f

### CODE REVIEW

**Comment by Siedlerchr** [comment]
To make it a bit clearer for the user, I would rename it "From ID based fetcher".
So the semantic workflow could be  "create new entry from isbn 12346567"

**Comment by oscargus** [comment]
Or just "From ID".

I think the idea is really good, but I'm thinking that there may be a better location, like a separate menu item. However, I assume that this is the dialog shown when pressing "new entry" in the toolbar (which I rarely use). No really good idea, but later it may be worthwhile trying to e.g. have a side panel with that functionality (or rather separating the fetcher types in search and ID, I realize).

File: src/main/resources/l10n/JabRef_sv.properties [hunk 1]
```diff
@@ -1640,3 +1640,8 @@ last_four_nonpunctuation_characters_should_be_numerals=
 Remember_password?=Kom_ihåg_lösenord?
 
 shared=delad
+Remember_password?=
```
Comment (oscargus) [inline src/main/resources/l10n/JabRef_sv.properties]: Duplicate. :-)

**Comment by tobiasdiez** [comment]
I'm also in favor of a new menu-item + dialog. This also provides you more room to include a bit more details. My proposal:
Image

Has the advantage that the user need not select the kind of ID he has (most of them are quite distinct and JabRef should be able to tell them apart) and he gets a preview of the entry (which tells him that we put the correct ID in)

**Comment by koppor** [comment]
@tobiasdiez I like the idea, but do it *in addition*. I like the current dialog by @zellerdev. I vote for doing both!

I like the current "ID-based entry generator" (@zellerdev please rename acoordingly). Rename "Search" to "Generate". Is it possible to place "Cancel" to the right?

File: src/main/java/net/sf/jabref/gui/EntryTypeDialog.java [hunk 2]
```diff
@@ -148,6 +174,95 @@ private JPanel createEntryGroupPanel(String groupTitle, Collection<EntryType> en
...
+                        JOptionPane.showMessageDialog(null, Localization.lang("Error_while_fetching_from_%0", fetcher.getName()), Localization.lang("Error_while_fetching"), JOptionPane.ERROR_MESSAGE);
+                    }
+                }
+                return bibEntry;
+            }
+
+            @Override
+            protected void done() {
+                if (isCancelled()) {
+                    JOptionPane.showMessageDialog(null, "Fetching was canceled", "Canceled fetching", JOptionPane.INFORMATION_MESSAGE);
```
Comment (koppor) [inline src/main/java/net/sf/jabref/gui/EntryTypeDialog.java]: Why is the mesasge not localized?

File: src/main/java/net/sf/jabref/gui/importer/fetcher/EntryFetchers.java [hunk 3]
```diff
@@ -30,13 +31,19 @@ public EntryFetchers(JournalAbbreviationLoader abbreviationLoader) {
...
         entryFetchers.add(new SearchBasedEntryFetcher(new ArXiv()));
         entryFetchers.add(new SearchBasedEntryFetcher(new GvkFetcher()));
     }
 
     public List<EntryFetcher> getEntryFetchers() {
         return Collections.unmodifiableList(this.entryFetchers);
     }
+
+    public static List<IdBasedFetcher> getIdFetchers() {
+        List<IdBasedFetcher> list = new LinkedList<>();
```
Comment (koppor) [inline src/main/java/net/sf/jabref/gui/importer/fetcher/EntryFetchers.java]: Why not `ArrayList`? It is [faster at sorting](http://stackoverflow.com/a/26013151/873282).

File: src/main/resources/l10n/JabRef_da.properties [hunk 4]
```diff
@@ -1781,3 +1780,8 @@ should_have_the_first_letter_capitalized=
 Remember_password?=
 
 shared=
+Finished_downloading_full_text_document_for_entry_%0.=
+
+No_entry_with_id_'%0'_for_fetcher_'%1'_was_found.=
+
+Error_while_fetching=Fejl_under_hentning
```
Comment (koppor) [inline src/main/resources/l10n/JabRef_da.properties]: Where do you know Dansk from? :innocent: - Are we really sure that there is no equivalent translation somewhere? What do the web-based fetchers output if something goes wrong?

File: src/main/java/net/sf/jabref/gui/EntryTypeDialog.java [hunk 5]
```diff
@@ -148,6 +174,95 @@ private JPanel createEntryGroupPanel(String groupTitle, Collection<EntryType> en
         return panel;
     }
 
+    private JPanel createIdFetcher(String groupTitle) {
```
Comment (koppor) [inline src/main/java/net/sf/jabref/gui/EntryTypeDialog.java]: Rename to "createIdFetcherPanel" and remove parameter. Just pass it directly to the method below and use Localization.lang to have the title translated into other languages.

**Comment by koppor** [comment]
Please add updated screenshot, also the German screenshot to be sure that everything is translated. Ater that, it should be good to go.

File: src/main/java/net/sf/jabref/gui/EntryTypeDialog.java [hunk 6]
```diff
@@ -9,33 +9,52 @@
...
 import net.sf.jabref.model.entry.BibtexEntryTypes;
 import net.sf.jabref.model.entry.EntryType;
 import net.sf.jabref.model.entry.IEEETranEntryTypes;
 
 import com.jgoodies.forms.builder.ButtonBarBuilder;
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
 import org.jdesktop.swingx.VerticalLayout;
 
+import static net.sf.jabref.gui.importer.fetcher.EntryFetchers.getIdFetchers;
```
Comment (tobiasdiez) [inline src/main/java/net/sf/jabref/gui/EntryTypeDialog.java]: I would prefer if you don't import this as static.

File: src/main/java/net/sf/jabref/gui/EntryTypeDialog.java [hunk 7]
```diff
@@ -91,12 +112,17 @@ private JPanel createEntryGroupsPanel() {
...
 
-        if(biblatexMode) {
+        if (biblatexMode) {
             panel.add(createEntryGroupPanel("BibLateX", EntryTypes.getAllValues(bibDatabaseContext.getMode())));
         } else {
             panel.add(createEntryGroupPanel("BibTeX", BibtexEntryTypes.ALL));
             panel.add(createEntryGroupPanel("IEEETran", IEEETranEntryTypes.ALL));
-            panel.add(createEntryGroupPanel(Localization.lang("Custom"), CustomEntryTypesManager.ALL));
+
+            if (!CustomEntryTypesManager.ALL.isEmpty()) {
```
Comment (tobiasdiez) [inline src/main/java/net/sf/jabref/gui/EntryTypeDialog.java]: The custom entry types and the id based entry are not displayed in BibLatexMode. Is this by design? @JabRef/developers

File: src/main/java/net/sf/jabref/gui/EntryTypeDialog.java [hunk 8]
```diff
@@ -91,12 +112,17 @@ private JPanel createEntryGroupsPanel() {
...
+                panel.add(createEntryGroupPanel(Localization.lang("Custom"), CustomEntryTypesManager.ALL));
+            }
+
+            panel.add(createIdFetcher("ID-based entry generator"));
```
Comment (tobiasdiez) [inline src/main/java/net/sf/jabref/gui/EntryTypeDialog.java]: -> not displayed in BibLatexMode?

File: src/main/java/net/sf/jabref/gui/EntryTypeDialog.java [hunk 9]
```diff
@@ -148,6 +174,95 @@ private JPanel createEntryGroupPanel(String groupTitle, Collection<EntryType> en
         return panel;
     }
 
+    private JPanel createIdFetcher(String groupTitle) {
+        JButton generateButton = new JButton(Localization.lang("Generate"));
+        JButton cancelButton = new JButton(Localization.lang("Cancel"));
+        JTextField idTextField = new JTextField("");
+        JComboBox<String> comboBox = new JComboBox<>();
+        getIdFetchers().forEach(n -> comboBox.addItem(n.getName()));
```
Comment (tobiasdiez) [inline src/main/java/net/sf/jabref/gui/EntryTypeDialog.java]: rename n -> fetcher

File: src/main/java/net/sf/jabref/gui/EntryTypeDialog.java [hunk 10]
```diff
@@ -148,6 +174,95 @@ private JPanel createEntryGroupPanel(String groupTitle, Collection<EntryType> en
...
+            protected Optional<BibEntry> doInBackground() throws Exception {
+                searchID = idTextField.getText();
+                if (!searchID.isEmpty()) {
+                    fetcher = EntryFetchers.getIdFetchers().get(comboBox.getSelectedIndex());
+                    try {
+                        bibEntry = fetcher.performSearchById(searchID);
+                        dispose();
+                    } catch (FetcherException e) {
+                        LOGGER.error("Error fetching from " + fetcher.getName(), e);
+                        JOptionPane.showMessageDialog(null, Localization.lang("Error_while_fetching_from_%0", fetcher.getName()), Localization.lang("Error_while_fetching"), JOptionPane.ERROR_MESSAGE);
```
Comment (tobiasdiez) [inline src/main/java/net/sf/jabref/gui/EntryTypeDialog.java]: without underscore

File: src/main/java/net/sf/jabref/gui/EntryTypeDialog.java [hunk 11]
```diff
@@ -148,6 +174,95 @@ private JPanel createEntryGroupPanel(String groupTitle, Collection<EntryType> en
...
+                    JOptionPane.showMessageDialog(null, "Fetching was canceled", "Canceled fetching", JOptionPane.INFORMATION_MESSAGE);
+                    return;
+                }
+
+                try {
+                    Optional<BibEntry> result = get();
+                    if (result.isPresent()) {
+                        frame.getCurrentBasePanel().insertEntry(result.get());
+                    } else {
+                        JOptionPane.showMessageDialog(null, Localization.lang("No_entry_with_id_'%0'_for_fetcher_'%1'_was_found.", searchID, fetcher.getName()), Localization.lang("No_files_found."), JOptionPane.WARNING_MESSAGE);
```
Comment (tobiasdiez) [inline src/main/java/net/sf/jabref/gui/EntryTypeDialog.java]: no underscore

File: src/main/java/net/sf/jabref/gui/EntryTypeDialog.java [hunk 12]
```diff
@@ -148,6 +174,95 @@ private JPanel createEntryGroupPanel(String groupTitle, Collection<EntryType> en
...
+                    }
+                } catch (ExecutionException | InterruptedException e) {
+                    e.printStackTrace();
```
Comment (tobiasdiez) [inline src/main/java/net/sf/jabref/gui/EntryTypeDialog.java]: Proper logging

File: src/main/java/net/sf/jabref/gui/EntryTypeDialog.java [hunk 13]
```diff
@@ -148,6 +174,95 @@ private JPanel createEntryGroupPanel(String groupTitle, Collection<EntryType> en
...
+                }
+            }
+        };
+
+        generateButton.addActionListener(n -> fetcherWorker.execute());
```
Comment (tobiasdiez) [inline src/main/java/net/sf/jabref/gui/EntryTypeDialog.java]: rename n

File: src/main/java/net/sf/jabref/gui/EntryTypeDialog.java (hunk 11 above)
Comment (tobiasdiez) [inline src/main/java/net/sf/jabref/gui/EntryTypeDialog.java]: fetcher could be null I think (select a fetcher in the dialog but let the id empty -> click search)

**Comment by tobiasdiez** [comment]
Just some small remarks, then this could be merged.

File: src/main/java/net/sf/jabref/gui/EntryTypeDialog.java (hunk 7 above)
Comment (Siedlerchr) [inline src/main/java/net/sf/jabref/gui/EntryTypeDialog.java]: No, they are already displayed (even in biblatex mode)  but they are mixed with the others:

jabrefmycustomentrytype

File: src/main/resources/l10n/JabRef_en.properties [hunk 14]
```diff
@@ -2177,7 +2177,7 @@ Copied_version_to_clipboard=Copied_version_to_clipboard
 
 BibTeX_key=BibTeX_key
 Message=Message
-
+ID-based_entry_generator=ID-based_entry_generator
```
Comment (koppor) [inline src/main/resources/l10n/JabRef_en.properties]: The order of entries in `JabRef_en.properties` is important. Please move this together with `No_entry_with_id_'%0'_for_fetche...`. Maybe at the end of the file (as it is common for new entries)

File: CHANGELOG.md [hunk 15]
```diff
@@ -19,6 +19,7 @@ We refer to [GitHub issues](https://github.com/JabRef/jabref/issues) by using `#
 - [#1897](https://github.com/JabRef/jabref/issues/1897) Implemented integrity check for `year` field: Last four nonpunctuation characters should be numerals
 - Address in MS-Office 2007 xml format is now imported as `location`
 - [#1912](https://github.com/JabRef/jabref/issues/1912) Implemented integrity check for `edition` field: Should have the first letter capitalized (BibTeX), Should contain an integer or a literal (BibLaTeX)
+- Add ID Fetcher in EntryTypeDialog
```
Comment (koppor) [inline CHANGELOG.md]: We decided for "ID-based_entry_generator". Thus, this term should be used here. Further, I don't believe it's called "EntryTypeDialog` on the user's side (see http://help.jabref.org/en/BaseFrameHelp). Maybe phrase it "Support entry generation based on IDs in the dialog for choosing the type for a new entry" or "The dialog for choosing new entries additionally supports ID-based entry generation"

File: src/main/java/net/sf/jabref/gui/EntryTypeDialog.java [hunk 16]
```diff
@@ -148,6 +172,95 @@ private JPanel createEntryGroupPanel(String groupTitle, Collection<EntryType> en
...
+                    }
+                }
+                dispose();
+                return bibEntry;
+            }
+
+            @Override
+            protected void done() {
+                if (isCancelled()) {
+                    JOptionPane.showMessageDialog(null, Localization.lang("%0 import canceled", fetcher.getName()),Localization.lang("Import canceled by user"), JOptionPane.INFORMATION_MESSAGE);
```
Comment (koppor) [inline src/main/java/net/sf/jabref/gui/EntryTypeDialog.java]: According to http://stackoverflow.com/a/6204600/873282 this is not thread safe as it is not called in the EDT. Maybe, it is also not called at all, according to the last comment by [searchengine27](http://stackoverflow.com/users/3560873/searchengine27) at http://stackoverflow.com/a/6204600/873282. Did you test whether this dialog appears?

File: src/main/java/net/sf/jabref/gui/EntryTypeDialog.java [hunk 17]
```diff
@@ -148,6 +172,95 @@ private JPanel createEntryGroupPanel(String groupTitle, Collection<EntryType> en
...
+
+                try {
+                    Optional<BibEntry> result = get();
+                    if (result.isPresent()) {
+                        frame.getCurrentBasePanel().insertEntry(result.get());
+                    } else {
+                        JOptionPane.showMessageDialog(null, Localization.lang("No entry with id '%0' for fetcher '%1' was found.", searchID, fetcher.getName()), Localization.lang("No files found."), JOptionPane.WARNING_MESSAGE);
+                    }
+                } catch (ExecutionException | InterruptedException e) {
+                    LOGGER.error("No entry with id '" + searchID + " for fetcher '" + fetcher.getName() + "' was found.");
```
Comment (koppor) [inline src/main/java/net/sf/jabref/gui/EntryTypeDialog.java]: We changed that when we set together. Why did you change that back? Is it because of some rebase stuff? In general, to not rebase, but do merge.

File: src/main/java/net/sf/jabref/gui/EntryTypeDialog.java [hunk 18]
```diff
@@ -148,6 +172,95 @@ private JPanel createEntryGroupPanel(String groupTitle, Collection<EntryType> en
...
+        constraints.gridy = 2;
+        JPanel buttons = new JPanel();
+        ButtonBarBuilder bb = new ButtonBarBuilder(buttons);
+        bb.addButton(generateButton);
+        bb.addButton(cancelButton);
+
+        jPanel.add(buttons, constraints);
+        jPanel.setBorder(BorderFactory.createTitledBorder(BorderFactory.createEtchedBorder(), Localization.lang("ID-based_entry_generator")));
+
+
```
Comment (koppor) [inline src/main/java/net/sf/jabref/gui/EntryTypeDialog.java]: Please remove line, so that **one** empty line is remaining.

File: src/main/resources/l10n/JabRef_de.properties [hunk 19]
```diff
@@ -2490,6 +2490,12 @@ last_four_nonpunctuation_characters_should_be_numerals=Die letzten vier Nichtint
 Remember_password?=Passwort_merken?
 
 shared=geteilt
+No_entry_with_id_'%0'_for_fetcher_'%1'_was_found.=Kein_Eintrag_mit_der_ID_'%0'_für_fetcher_'%1'_wurde_gefunden.
```
Comment (koppor) [inline src/main/resources/l10n/JabRef_de.properties]: We also updated this when we met. Why is our new String lost?

**Comment by koppor** [comment]
Can you rewrite the code to depend on #1885 only. As #1923 depends on #1929, it will take more than this week to get it merged. I don't want to wait that long.
